File: typescript/Behavioral/20-Strategy/Strategy/HourlyRateStrategy.ts
import { IPricingStrategy } from "./IPricingStrategy";
import type { JobCard } from "../Model/JobCard";

export class HourlyRateStrategy implements IPricingStrategy {
  readonly name = "Hourly Rate";

  constructor(
    private readonly hourlyRate: number = 450,
    private readonly minimumCallout: number = 350,
  ) {}

  calculate(job: JobCard): number {
    return Math.max(job.hoursWorked * this.hourlyRate, this.minimumCallout);
  }

  explain(job: JobCard): string {
    const raw = job.hoursWorked * this.hourlyRate;

    return raw < this.minimumCallout
      ? `${job.hoursWorked}h × N$${this.hourlyRate.toFixed(2)} = N$${raw.toFixed(2)}, ` +
          `below minimum callout — charged N$${this.minimumCallout.toFixed(2)}.`
      : `${job.hoursWorked}h × N$${this.hourlyRate.toFixed(2)} = N$${raw.toFixed(2)}.`;
  }
}

File: csharp/Behavioral/20-Strategy/Strategy/HourlyRateStrategy.cs
namespace Strategy;

public class HourlyRateStrategy : IPricingStrategy
{
    private readonly decimal _hourlyRate;
    private readonly decimal _minimumCallout;

    public HourlyRateStrategy(decimal hourlyRate = 450m, decimal minimumCallout = 350m)
    {
        _hourlyRate     = hourlyRate;
        _minimumCallout = minimumCallout;
    }

    public string Name => "Hourly Rate";

    public decimal Calculate(JobCard job) =>
        Math.Max(job.HoursWorked * _hourlyRate, _minimumCallout);

    public string Explain(JobCard job)
    {
        var raw = job.HoursWorked * _hourlyRate;

        return raw < _minimumCallout
            ? $"{job.HoursWorked}h × N${_hourlyRate:N2} = N${raw:N2}, " +
              $"below minimum callout — charged N${_minimumCallout:N2}."
            : $"{job.HoursWorked}h × N${_hourlyRate:N2} = N${raw:N2}.";
    }
}
